### PR TITLE
Include catalog realms under Community Catalogs

### PR DESCRIPTION
**Pull request by FadhlanR**
In this PR, I introduced a new API in the realm server called `_public-realms`. This API is used to display catalog realms under the "Community Catalogs" section in the workspace chooser.

https://github.com/user-attachments/assets/05072d90-ec26-4476-bf6b-c9ab3a002c42

### CODE REVIEW

File: packages/host/app/components/operator-mode/workspace-chooser/index.gts [hunk 1]
```diff
@@ -12,17 +16,44 @@ interface Signature {
...
+
+  private get displayCatalogWorkspaces() {
+    return this.catalogWorkspaceURLs && this.catalogWorkspaceURLs.length > 0;
+  }
+
+  private get catalogWorkspaceURLs() {
+    return this.fetchCatalogRealmURLs.value;
+  }
+
+  private fetchCatalogRealmURLs = trackedFunction(this, async () => {
```
Comment (jurgenwerk) [inline packages/host/app/components/operator-mode/workspace-chooser/index.gts]: The other day I discovered `ember-resources` also has `trackedTask` which can be put around a `restartableTask` for example, and then it's easy to get the benefits of an ember concurrency task (cancellation when component is torn down for example), and you get a reactive property for when the task is in progress (fetching is happening), using `{{trackedTask.isRunning}}` if you needed to show a fetching state, a spinner or something. Just saying this for awareness–next time we do some simple fetching we could consider that.

File: packages/realm-server/tests/realm-server-test.ts [hunk 2]
```diff
@@ -3012,6 +3012,27 @@ module('Realm Server', function (hooks) {
         ],
       });
     });
+
+    test('can fetch public realms', async function (assert) {
```
Comment (jurgenwerk) [inline packages/realm-server/tests/realm-server-test.ts]: Would be good to add a test for when one of the realm info fetches fails during executing this request

File: packages/host/app/components/operator-mode/workspace-chooser/index.gts (hunk 1 above)
Comment (lukemelia) [reply packages/host/app/components/operator-mode/workspace-chooser/index.gts]: ember-concurrency task derived state is already reactive -- you wouldn't need something in addition for that.

File: packages/host/app/services/realm.ts [hunk 3]
```diff
@@ -371,6 +379,25 @@ export default class RealmService extends Service {
     return resource.token;
   }
 
+  async fetchPublicRealmURLs(): Promise<string[]> {
+    let realmServerURL = new URL(window.location.href);
+    if (ENV.environment === 'development' || ENV.environment === 'test') {
+      realmServerURL = new URL('http://localhost:4201');
```
Comment (lukemelia) [inline packages/host/app/services/realm.ts]: This kind of conditional is starting to appear in different parts of the code base. We need one place to deal with it (perhaps the realm-server service?)

File: packages/realm-server/scripts/start-development.sh [hunk 4]
```diff
@@ -15,6 +15,7 @@ NODE_ENV=development \
   REALM_SECRET_SEED="shhh! it's a secret" \
   MATRIX_URL=http://localhost:8008 \
   REALM_SERVER_MATRIX_USERNAME=realm_server \
+  EXCLUDED_PUBLIC_REALMS='http://localhost:4201/seed/,https://cardstack.com/base/,http://localhost:4202/test/,http://localhost:4202/node-test/,http://localhost:4203/,http://localhost:4204/,http://localhost:4205/test/' \
```
Comment (lukemelia) [inline packages/realm-server/scripts/start-development.sh]: It would be good to come up with a better way to manage this. Perhaps there should be a flag that can be set in realm.json?

File: packages/realm-server/tests/realm-server-test.ts (hunk 2 above)
Comment (FadhlanR) [reply packages/realm-server/tests/realm-server-test.ts]: Not sure how to do that in realm server test. Is there a way to mock an API like `_info` so we can make the response failed?

File: packages/host/app/services/realm.ts (hunk 3 above)
Comment (habdelra) [reply packages/host/app/services/realm.ts]: it is actually a property available in the new realm-server service in my PR

File: packages/realm-server/tests/realm-server-test.ts (hunk 2 above)
Comment (jurgenwerk) [reply packages/realm-server/tests/realm-server-test.ts]: I think you can mount a handler in virtual network that is passed to the server and to the realm, such that it throws an error if it encounters a specific url. Perhaps something like this:

image

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -101,6 +101,12 @@ jobs:
       - name: Wait for ember-cli to be ready
         run: pnpm npx wait-for-localhost 4200
         working-directory: packages/host
+      - name: Start matrix
+        run: pnpm start:synapse
+        working-directory: packages/matrix
+      - name: Initialize matrix data
+        run: pnpm drop-all-dbs
+        working-directory: packages/realm-server
```
Comment (habdelra) [inline .github/workflows/ci.yaml]: I'm curious why you needed to add this, was something bleeding into these tests? Ideally we should not have to drop all the DB's since there are none to begin with, and the matrix server state is cleared in between each test. 

Perhaps what this really means is that you need the isolated realm for your tests like we use for the registration tests.

File: packages/realm-server/migrations/1727094475141_add-public-realms-index.js.js [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+      break;
+    case 'production':
+      pgm.sql(`DELETE FROM realm_user_permissions WHERE realm_url = 'https://app.boxel.ai/experiments/' AND username = '*' AND read = true`);
+      break;
+    default:
+      pgm.sql(`DELETE FROM realm_user_permissions WHERE realm_url = 'http://localhost:4201/experiments/' AND username = '*' AND read = true`);
+      pgm.sql(`DELETE FROM realm_user_permissions WHERE realm_url = 'http://localhost:4204/' AND username = '*' AND read = true`);
+      break;
+  }
+};
```
Comment (habdelra) [inline packages/realm-server/migrations/1727094475141_add-public-realms-index.js.js:15]: make sure to provide an exports.down too where you add these permissions back

File: packages/realm-server/server.ts [hunk 7]
```diff
@@ -601,6 +603,51 @@ export class RealmServer {
...
   }
+
+  private fetchCatalogRealms(): (
+    ctxt: Koa.Context,
+    _next: Koa.Next,
+  ) => Promise<any> {
+    return async (ctxt: Koa.Context, _next: Koa.Next) => {
+      let results = await fetchPublicRealms(this.dbAdapter);
+      let promises = results.map(async ({ realm_url: realmURL }) => {
+        let response = await this.virtualNetwork.handle(
```
Comment (habdelra) [inline packages/realm-server/server.ts]: this variable `response` is confusing because its easy to consuse this with the HTTP response for this endpoint. can we rename it like `realmInfoResponse`?

File: packages/realm-server/server.ts [hunk 8]
```diff
@@ -601,6 +603,51 @@ export class RealmServer {
...
     throw new Error(`Can not determine the matrix registration secret`);
   }
+
+  private fetchCatalogRealms(): (
+    ctxt: Koa.Context,
+    _next: Koa.Next,
+  ) => Promise<any> {
+    return async (ctxt: Koa.Context, _next: Koa.Next) => {
+      let results = await fetchPublicRealms(this.dbAdapter);
+      let promises = results.map(async ({ realm_url: realmURL }) => {
```
Comment (habdelra) [inline packages/realm-server/server.ts]: lets rename `promises` to `publicRealmPromises`. optionally you could just wrap line 642's `Promise.all` around this and not even need a `promises` variable

File: .github/workflows/ci.yaml (hunk 5 above)
Comment (FadhlanR) [reply .github/workflows/ci.yaml]: Based on a discussion with @lukemelia , we decided to make the experiments realm not publicly readable. After making those changes, all matrix tests failed. I noticed this happened because the indexing failed, which was due to not having realm users when executing `pnpm start:without-matrix`. I believe I don’t need to use `drop-all-dbs`; instead, I can use `register-all`.

File: packages/realm-server/server.ts [hunk 9]
```diff
@@ -601,6 +603,51 @@ export class RealmServer {
...
+          return;
+        }
+
+        return {
+          type: 'catalog-realm',
+          id: realmURL,
+          attributes: json.data.attributes,
+        };
+      });
+      let data = (await Promise.all(promises)).filter(Boolean);
```
Comment (habdelra) [inline packages/realm-server/server.ts]: it would be nice to cache this answer. right now its a big huge ordeal to make a new public catalog. it's not possible that one will be made on the fly. So let's just calculate this answer once and cache it. In the future when it's possible to make public catalogs on the fly we can revisit this.

File: .github/workflows/ci.yaml (hunk 5 above)
Comment (habdelra) [reply .github/workflows/ci.yaml]: hmmmm so this is really annoying because the matrix tests never touch the experiments realm. and honestly we only really need to start the base realm--not all those other realms as part of `start:all`. running register-all is fine. but I think a better solution would be to refactor `pnpm start:withou-matrix` to only start the base realm, and not start any of the other realms. probably in that case we'd want to rename the script to something more precise, like `pnpm start:services-for-matrix-tests`.

File: packages/runtime-common/realm-index-updater.ts
Comment (habdelra) [inline packages/runtime-common/realm-index-updater.ts:1]: I'm curious about these changes--presumably this is the same as the `main` branch I wonder why it is marked as a change? perhaps you need to merge upstream changes?

File: packages/runtime-common/realm.ts [hunk 10]
```diff
@@ -574,7 +575,12 @@ export class Realm {
 
   async #startup() {
     await Promise.resolve();
-    await this.#realmIndexUpdater.run();
+    let isNewIndex = await this.#realmIndexUpdater.isNewIndex();
+    let promise = this.#realmIndexUpdater.run();
+    if (isNewIndex) {
+      // we only await the full indexing at boot if this is a brand new index
+      await promise;
+    }
```
Comment (habdelra) [inline packages/runtime-common/realm.ts]: same comment as above--why would this be marked as a diff? perhaps you need to merge upstream?

**Comment by FadhlanR** [comment]
This PR is ready for another review. I made a couple of changes:

- I updated the experiment realm's permissions so it is no longer publicly readable, which prevents it from being displayed in the `Community Catalog` section.
- I added a new property called `showAsCatalog` in the `realm.json` file. This property is used to determine whether public realms will be displayed as catalog realms. Consequently, I renamed the new API from `_public-realms` to `_catalog-realms` to clarify that this API won’t return all public realms.
- Previously, we stored realm URLs in `unresolvedRealmURLs` in `CardService`, and the value of that property came from matrix account data. Since implementing the catalog realms API, I moved the realm URL storage to `RealmServerService`, so now realm URLs from both account data and the catalog realms API are stored in one place.
- Following @jurgenwerk 's feedback, I added a new test to simulate a scenario if realm info is failed to fetch.

File: .github/workflows/ci.yaml (hunk 5 above)
Comment (habdelra) [reply .github/workflows/ci.yaml]: if you opt to use `register-all` could you make a ticket to refactor `pnpm start:without-matrix` to not start unnecessary realms. then we won't loose track of the cleanup work

Comment (FadhlanR) [reply .github/workflows/ci.yaml]: I'll implement the `start:services-for-matrix-tests` script. Previously, we used the experiments realm in the matrix tests, but in this PR, I've removed it, so I believe that script can now be implemented.

**Review by lukemelia — APPROVED** [review]

File: packages/host/app/services/realm-server.ts [hunk 11]
```diff
@@ -87,6 +91,69 @@ export default class RealmServerService extends Service {
...
+    }
+
+    await this.fetchCatalogRealmURLs();
+  }
+
+  private async fetchCatalogRealmURLs() {
+    let realmServerURL = new URL(window.location.href);
+    if (
+      realmServerURL.origin === 'http://localhost:4200' ||
+      realmServerURL.origin === 'http://localhost:7357'
```
Comment (jurgenwerk) [inline packages/host/app/services/realm-server.ts]: Where is 7357 coming from?

File: packages/realm-server/package.json [hunk 12]
```diff
@@ -89,6 +89,7 @@
     "start:staging": "./scripts/start-staging.sh",
     "start:development": "./scripts/start-development.sh",
     "start:production": "./scripts/start-production.sh",
+    "start:services-for-matrix-tests": "./scripts/start-services-for-matrix-tests.sh",
```
Comment (jurgenwerk) [inline packages/realm-server/package.json:92]: Is this now preferable over `start-without-matrix` to run matrix tests?

File: packages/host/app/services/realm-server.ts (hunk 11 above)
Comment (habdelra) [reply packages/host/app/services/realm-server.ts]: i believe 7357 is the port that ember CLI runs on in CI. But we should really be using `this.url` here

File: packages/host/app/services/realm-server.ts [hunk 13]
```diff
@@ -87,6 +91,69 @@ export default class RealmServerService extends Service {
...
+
+    if (!this._userRealmURLs.includes(baseRealmUrl)) {
+      this._userRealmURLs.unshift(baseRealmUrl);
+    }
+
+    await this.fetchCatalogRealmURLs();
+  }
+
+  private async fetchCatalogRealmURLs() {
+    let realmServerURL = new URL(window.location.href);
```
Comment (habdelra) [inline packages/host/app/services/realm-server.ts]: use `this.url` instead

File: packages/realm-server/package.json (hunk 12 above)
Comment (habdelra) [reply packages/realm-server/package.json:92]: yeah, it's gonna run only the realms necessary for matrix tests. you can still use `start:all` if you want--but you'll be running realms you dont need (and you'll have to turn off synapse, since playwright needs to control that docker container).

File: packages/realm-server/scripts/start-services-for-matrix-tests.sh [hunk 14]
```diff
@@ -0,0 +1,7 @@
+#! /bin/sh
+NODE_NO_WARNINGS=1 start-server-and-test \
+  'run-p start:pg start:base start:base:root' \
+  'http-get://localhost:4201/base/_readiness-check?acceptHeader=application%2Fvnd.api%2Bjson' \
+  'run-p start:test-realms' \
+  'http-get://localhost:4202/test/_readiness-check?acceptHeader=application%2Fvnd.api%2Bjson|http-get://localhost:4202/node-test/_readiness-check?acceptHeader=application%2Fvnd.api%2Bjson' \
+  'wait'
```
Comment (habdelra) [inline packages/realm-server/scripts/start-services-for-matrix-tests.sh:7]: :+1:

**Review by habdelra — APPROVED** [review]
Just a minor note around not duplicating the logic of `RealmServerService.url`. aside from that it looks good.
